### PR TITLE
[fix] non numeric task ids crash concent api

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/concent/issues/55
https://www.pivotaltracker.com/story/show/154444926

### CODE REVIEW

File: concent_api/core/migrations/0005_auto_20180122_0044.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    dependencies = [
+        ('core', '0004_auto_20180119_1055'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='message',
+            name='task_id',
+            field=models.CharField(max_length=255, null=True),
+        ),
```
Comment (cameel) [inline concent_api/core/migrations/0005_auto_20180122_0044.py]: This change requires a data migration to preserve existing data.

File: concent_api/core/models.py [hunk 2]
```diff
@@ -11,7 +11,7 @@ class Message(Model):
     type        = PositiveSmallIntegerField()
     timestamp   = DateTimeField()
     data        = BinaryField()
-    task_id     = IntegerField(null = True)
+    task_id     = CharField(max_length = 255, null = True)
```
Comment (cameel) [inline concent_api/core/models.py]: Could you check if Golem code has any limit on the length of task IDs?

I think that if we impose a limit here, we should have some kind of validation of that limit in golem-messages. And perhaps a constant that defined the maximum ID length.

File: concent_api/core/tests/test_core_views.py [hunk 3]
```diff
@@ -129,6 +129,41 @@ def test_send_should_return_http_200_if_message_timeout(self):
         )
         self.assertIsInstance(response_message, GolemMessage)
 
+    @freeze_time("2017-11-17 10:00:00")
+    def test_send_should_accept_valid_message_with_non_numeric_task_id(self):
```
Comment (cameel) [inline concent_api/core/tests/test_core_views.py]: Could you extend this test to have a list of valid and invalid ids and check them in a loop?

File: concent_api/core/migrations/0005_auto_20180122_0044.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/core/migrations/0005_auto_20180122_0044.py]: It doesn't, int will be casted to string without problems.

File: concent_api/core/models.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/core/models.py]: I have checked in docs and code and couldn't find any.

File: concent_api/core/tests/test_core_views.py (hunk 3 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: What would be invalid id ?

File: concent_api/core/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/core/models.py]: How this can break support for non-numeric IDs. Have you meant numeric IDs ? Wasn't that the purpose of this change ?

**Comment by rwrzesien** [comment]
> It seems to me that the the code is going to fail and return a 5xx error if you try to use a non-numeric ID. This is not acceptable. You should add a validation to the validating function in views. The view should return a 4xx error if a message is invalid.

There is a test verifying that, could you please provide an example of id that will break ?

**Comment by cameel** [comment]
Try `1`.

**Comment by rwrzesien** [comment]
> Try 1.
We have a lot of tests which use exactly this ID and they all pass.

File: concent_api/core/migrations/0005_auto_20180122_0044.py (hunk 1 above)
Comment (cameel) [reply concent_api/core/migrations/0005_auto_20180122_0044.py]: Interesting. So PostgreSQL can do int <-> string conversion in simple cases. That's cool.

File: concent_api/core/models.py (hunk 2 above)
Comment (cameel) [reply concent_api/core/models.py]: OK. But we still need a validation.

File: concent_api/core/migrations/0005_auto_20180122_0044.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/core/migrations/0005_auto_20180122_0044.py]: Int => string yes
String => int probably will ask about custom handling while migrating, if you don't provide any and there will be non-numeric string then it will fail.

File: concent_api/core/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cameel) [reply concent_api/core/models.py]: >  Have you meant numeric IDs ?

Yes, sorry for the confusion.

> Wasn't that the purpose of this change ?

The purpose was to add support for non-numeric IDs. Your change does this but at the same time breaks numeric IDs in a way.

Looks like Golem lets you use anything as ID. And I did not see any string conversions there so I suspect that to Golem ids like `666` and `'666'` are different things since that's how Python works.

After your change we "support" both of these in the sense that there's no exception. But Django converts anything you assign to a `CharField` to string and `filter()` will find `'666'` if you give it `666`. So we don't really support numeric IDs any longer. We just convert them to strings (which Golem does not do).

**Comment by cameel** [comment]
> We have a lot of tests which use exactly this ID and they all pass.

OK, so it does not raise an exception. That's even worse because it's still broken - at least as long as we support both numeric and non-numeric IDs - because I think that to Golem `1` and `'1'` are different IDs.

I did not know that Django will convert anything you assign to a `CharField` to a string. My natural assumption was that you'd get a validation error or a type error on save. That's troubling and I think we'll have a lot of problems with silent conversions like this. Golem does not do these conversions and we can end up with slightly different data in messages and in the database. That can lead to bugs that are hard to find unless you're very careful.

I think that the only way out of this is to agree with Golem on a more strict schema for messages. At least that IDs must always be strings and nothing else. Then on our side validate this assumption (rather than blindly convert them). There should be a validation for this in golem-messages too.

File: concent_api/core/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/core/models.py]: Right, but I thought that is how we want it to work. So lets wait for Darek's response to use only strings which I think is best solution in that case, otherwise we would have to store extra info on the original type or use separate field for ints and strings. 
Btw. I don't see any casting of task_id in golem_messages, is there any option, apart from asking Darek, which you already did, to check how it works on Golem side ?

**Comment by rwrzesien** [comment]
Ok, I have added validation which assumes only strings, however I will wait with adding tests until Darek will respond.

File: concent_api/core/tests/test_core_views.py (hunk 3 above)
Comment (cameel) [reply concent_api/core/tests/test_core_views.py]: So to Golem there are no invalid IDs... that's bad because I think that using stuff like `[]`, `{}` `None`, `b'abc'`, `(1, 2, 3)`, `ForceReportComputedTask()` as IDs should not be allowed. It can easily lead to unexpected behavior if you don't carefully think about every possible corner case.

I think that you should at least add a test that checks these values. And you should test both `send/` and `receive/`.

Comment (rwrzesien) [reply concent_api/core/tests/test_core_views.py]: I will add them as soon as we clarify with Golem team to use only strings.

**Comment by jiivan** [comment]
@rwrzesien @cameel I think the discussion got a little bit blurred. I've answered in https://github.com/golemfactory/concent/issues/55#issuecomment-359509983
It's a string and will be validated against 128char length in https://github.com/golemfactory/golem-messages/issues/99

**Comment by rwrzesien** [comment]
@cameel I have added test for incorrect values, correct values are pretty much tested in other tests, which I have also needed to update. We don't currently do any validation in receive because we don't store any new message there.

File: concent_api/core/migrations/0005_auto_20180122_0044.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('core', '0004_auto_20180119_1055'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='message',
+            name='task_id',
+            field=models.CharField(max_length=255, null=True),
```
Comment (cameel) [inline concent_api/core/migrations/0005_auto_20180122_0044.py]: So this should now be 128. Please regenerate the migration from scratch.

Comment (cameel) [reply concent_api/core/migrations/0005_auto_20180122_0044.py]: Also, string fields in Django should be non-nullable.

File: concent_api/concent_api/settings/base.py [hunk 5]
```diff
@@ -209,3 +209,6 @@
 # - "Gatekeeper":  An internal helper that validates file transfer tokens.
 # - "admin-panel": Django admin panel that provides access to database content and service statistics.
 # CONCENT_FEATURES = []
+
+# Defines max length of task_id passed in Golem Messages.
+CONCENT_MAX_TASK_ID_LENGTH = 128
```
Comment (cameel) [inline concent_api/concent_api/settings/base.py]: Oh, one more thing. This should not be a setting. We don't really support user changing this arbitrarily - we need it to match golem-messages. It should be a constant. Please put it in models.

File: concent_api/core/models.py [hunk 6]
```diff
@@ -11,7 +12,7 @@ class Message(Model):
     type        = PositiveSmallIntegerField()
     timestamp   = DateTimeField()
     data        = BinaryField()
-    task_id     = IntegerField(null = True)
+    task_id     = CharField(max_length = settings.CONCENT_MAX_TASK_ID_LENGTH, null = True)
```
Comment (cameel) [inline concent_api/core/models.py]: String fields should not be non-nullable.

**Comment by rwrzesien** [comment]
@cameel Updated and rebased on master.

File: concent_api/core/models.py [hunk 7]
```diff
@@ -1,17 +1,19 @@
...
 
+from .constants       import MESSAGE_TASK_ID_MAX_LENGTH
+
 
 class Message(Model):
     type        = PositiveSmallIntegerField()
     timestamp   = DateTimeField()
     data        = BinaryField()
-    task_id     = IntegerField(null = True)
+    task_id     = CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH, blank = True, default = '')
```
Comment (cameel) [inline concent_api/core/models.py]: Please remove the default. The field should be assigned a value explicitly.

Also, now I see that this field should really be `blank = False` and `unique` too.

**Comment by rwrzesien** [comment]
@cameel Updated and rebased on master.

File: concent_api/core/migrations/0005_auto_20180129_1639.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    dependencies = [
+        ('core', '0004_auto_20180119_1055'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='message',
+            name='task_id',
+            field=models.CharField(default='1517243982.234565', max_length=128),
```
Comment (cameel) [inline concent_api/core/migrations/0005_auto_20180129_1639.py]: WTF?
